### PR TITLE
Improve TypeScript types with support for arrays containing elements of multiple types

### LINKED ISSUE #12: Variadic types
Today, the TS definition does not support having different type for each promise passed.

This is a known limitation in Typescript and there is a funny workaround that can be seen in the definition for `Promise.all()`: https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts#L113

Would you accept a PR that would apply the same pattern in the `d.ts` file?

### PR DESCRIPTION
**Pull request by ValentinH**
Closes #12

I used the approach used in the following PR to type `Promise.all()`: https://github.com/microsoft/TypeScript/pull/33707/files

### CODE REVIEW

**Comment by ValentinH** [comment]
Linting is also failing on `master`. I'm trying to find a solution.

**Comment by ValentinH** [comment]
I've updated `xo` to fix the tests.

**Comment by ValentinH** [comment]
@sindresorhus could you please have a look 😊

File: index.d.ts [hunk 1]
```diff
@@ -17,6 +17,8 @@ declare namespace pSettle {
 	type PromiseResult<ValueType> = pReflect.PromiseResult<ValueType>;
 	type PromiseFulfilledResult<ValueType> = pReflect.PromiseFulfilledResult<ValueType>;
 	type PromiseRejectedResult = pReflect.PromiseRejectedResult;
+
+	type Awaited<T> = T extends undefined ? T : T extends PromiseLike<infer U> ? U : T;
```
Comment (sindresorhus) [inline index.d.ts:21]: Can you add a TODO comment to remove this when TS supports this natively (which it will in TS 4).

Comment (ValentinH) [reply index.d.ts:22]: Done

**Pull request merged**